### PR TITLE
Add MongoDatabaseAdapter

### PR DESCRIPTION
**Pull request by dimas-b**
Using naive keyPrefix handling for now.

### CODE REVIEW

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return "MongoDB";
+  }
+
+  @Override
+  public Builder<MongoDatabaseAdapterConfig> newBuilder() {
+    return new Builder<MongoDatabaseAdapterConfig>() {
+      @Override
+      protected MongoDatabaseAdapterConfig getDefaultConfig() {
+        return ImmutableMongoDatabaseAdapterConfig.builder()
+            .connectionString("unknown")
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java:34]: why do we need those `unknown`s?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+  protected MongoDatabaseAdapter(MongoDatabaseAdapterConfig config) {
+    super(config);
+
+    keyPrefix = ByteString.copyFromUtf8(config.getKeyPrefix());
+    globalPointerKey = keyPrefix.toByteArray();
+
+    ConnectionString cs = new ConnectionString(config.getConnectionString());
+    MongoClientSettings settings = MongoClientSettings.builder().applyConnectionString(cs).build();
+
+    mongoClient = MongoClients.create(settings);
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:87]: `MongoDatabaseAdapterConfig` needs a functionality to provide a `MongoClient` instance to `MongoDatabaseAdapterConfig` (like `RocksDatabaseAdapter` does). I.e. if the `MongoDatabaseAdapterConfig` provides a non-null one, use that, otherwise create one. Probably better to wrap that `MongoClient` in a class similar to `RocksDbInstance`, which would also care about inquiring the collections (the `database.getCollection()` calls / keep the `MongoCollection` instances.

In other words: this constructor should be fast and cheap, if the MongoClient is provided via the config.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+public class MongoDatabaseAdapter
+    extends NonTransactionalDatabaseAdapter<MongoDatabaseAdapterConfig> {
+
+  private static final String GLOBAL_POINTER = "global_pointer";
+  private static final String GLOBAL_LOG = "global_log";
+  private static final String COMMIT_LOG = "commit_log";
+  private static final String KEY_LIST = "key_list";
+
+  private final ByteString keyPrefix;
+  private final byte[] globalPointerKey;
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:70]: Can we make this a `String`?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+
+    super.initializeRepo(defaultBranchName);
+  }
+
+  @Override
+  public void close() {
+    mongoClient.close();
+  }
+
+  private byte[] toId(Hash id) {
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:110]: Represent hashes as Strings in Mongo?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 3 above)
Comment (harshm-dremio) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:70]: possibly it's because of a security concern, since strings stay in memory, whereas arrays are cleared once code is done with it.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java (hunk 1 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java:34]: I did not want to make the optional in the "immutable" config since they are actually required for any connection.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 2 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:87]: I see. Did not realize this was the intention :)

However, why would we pass proper "configuration" settings to this class in this case? Why not always pass the "client" / "instance" object and let the factory configure it?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 3 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:70]: We could use `String` of course, but it looks like all the `persist.proto` objects convert to byte arrays naturally (including keys). So I figured we could use a byte array for the global pointer too.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 4 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:110]: Sure, but I think it's going to be more processing work with the current `persist.proto` objects :thinking: I'll give it a try.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 3 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:70]: It's not about the `byte[]` per se, it's to represent the `globalPointerKey` in a say "mongo db type friedly" way.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 4 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:110]: It's really just about representing it using the types available in mongo. (note: the jdbc-like adapters use `VARCHAR` as well).

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java (hunk 1 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java:34]: not if the connection-instance is passed ;)

Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java:34]: Yeah, but then we have a logical conflict between different ways of defining the connection in the same config object, don't we?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 4 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:110]: Mongo uses binary serialization internally, AFAIK, and supports byte arrays, but I see the rationale for using a human-readable type for keys :+1:

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 3 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: Reworked to `String` in 1f3f1e4c3afa0f1266f7906e09a58fb549318f66

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 4 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: Reworked to `String` in 1f3f1e4c3afa0f1266f7906e09a58fb549318f66

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 2 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:86]: I'm thinking about splitting "config" and "database instance" method at the factoring level :hourglass:

Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: I took an approach similar to `RocksDbInstance` in cc90e03fe60a625171481980ca86e887f73a7abe for now.

Still thinking about refactoring the config/factoring classes :thinking:

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import org.projectnessie.versioned.persist.adapter.DatabaseAdapterConfig;
+
+@Value.Immutable(lazyhash = true)
+public interface MongoDatabaseAdapterConfig extends DatabaseAdapterConfig {
+
+  String getConnectionString();
+
+  String getDatabaseName();
+
+  @Value.Derived
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:28]: I suspect this should be `@Value.Default` ("derived" means it's always built, "default" means its built when not explicitly set)

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+    return loadById(client.getGlobalLog(), id, GlobalStateLogEntry::parseFrom);
+  }
+
+  @Override
+  protected List<GlobalStateLogEntry> fetchPageFromGlobalLog(
+      NonTransactionalOperationContext ctx, List<Hash> hashes) {
+    return fetchPage(client.getGlobalLog(), hashes, GlobalStateLogEntry::parseFrom);
+  }
+
+  @FunctionalInterface
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:397]: Can you move this interface to `ProtoSerialization` so we can reuse it?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+    byte[] expectedGlobalId = expected.getGlobalId().toByteArray();
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
+                Filters.and(Filters.eq(globalPointerKey), Filters.eq("globalId", expectedGlobalId)),
+                new Document("$set", doc));
+
+    return result.wasAcknowledged()
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:364]: This CAS-operation LGTM from my understanding of MongoDB. But would be good if someone else could confirm.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+  }
+
+  @Override
+  protected CommitLogEntry fetchFromCommitLog(NonTransactionalOperationContext ctx, Hash hash) {
+    return loadById(client.getCommitLog(), hash, ProtoSerialization::protoToCommitLogEntry);
+  }
+
+  private Hash idAsHash(Document doc) {
+    String id = doc.get("_id", String.class);
+    String hash = id.substring(keyPrefix.length() + 1); // + 1 for the separator char
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:256]: Better use String.indexOf() here instead of assuming the same key-prefix (unlikely that this might happen)

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+   * Fetches a collection of documents by hash and returns them in the order of hashes requested.
+   *
+   * <p>Note: unknown hashes will correspond to {@code null} elements in the result list.
+   */
+  private <T> List<T> fetchMappedPage(
+      MongoCollection<Document> collection, List<Hash> hashes, Function<Document, T> mapper) {
+    List<String> ids = hashes.stream().map(this::toId).collect(Collectors.toList());
+    FindIterable<Document> docs = collection.find(Filters.in("_id", ids)).limit(hashes.size());
+
+    HashMap<Hash, Document> loaded = new HashMap<>(hashes.size());
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:215]: The initial-capacity is wrong here. You need `hashes.size() * 4 / 3 + 1` here.

**Review by snazy — COMMENTED** [review]
Mostly LGTM, few minor things

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java (hunk 5 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:28]: That's what I meant actually. `Derived` -> built on first get and cached. Since the constructor is trivial, I thought it'd make sense instead of forcing it to be created and set from the outside. WDYT?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 9 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: Good point. Fixed in e6b4f6eebea447420690ab610184702adf28f57c

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 8 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:271]: I've added some validation in e6b4f6eebea447420690ab610184702adf28f57c

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 6 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: Moved: dbf2425824c655c50cc7d19bc8df6fe742afff59

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java (hunk 5 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:28]: Immutables takes care of that piece in the generated code.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 8 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:271]: That's a lot ;)
I'm okay with that.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java (hunk 5 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:28]: Yes, so "derived" is created only on first access. Why is "default" preferable?

Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java]: Made `Client` settable in d7a1e320814af38cdfd0b5d4c75487e841fd7de9

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import org.immutables.value.Value;
+import org.projectnessie.versioned.persist.adapter.DatabaseAdapterConfig;
+
+@Value.Immutable(lazyhash = true)
+public interface MongoDatabaseAdapterConfig extends DatabaseAdapterConfig {
+
+  String getConnectionString();
+
+  String getDatabaseName();
+
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:28]: Just realized: need `withDatabaseName` + `withConnectionString` here as well.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java (hunk 1 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java:34]: It having these defaults would prevent checks in `MongoDatabaseClient` whether the properties are set though, right?

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+      NonTransactionalOperationContext ctx,
+      GlobalStatePointer expected,
+      GlobalStatePointer newPointer) {
+    Document doc = toDoc(newPointer);
+    byte[] expectedGlobalId = expected.getGlobalId().toByteArray();
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
```
Comment (rymurr) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:379]: How has this been tested? From my reading of the [docs](https://docs.mongodb.com/manual/reference/method/db.collection.updateOne/) the `updateOne` method in this way is the best/correct operator. However most google searches about conditional updates talk about pipeline operators which makes me wonder if its possible that the filter in the first argument returns an entry then the 2nd argument modifies it but these two operations aren't done atomically and there could be a race under high load.

Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:379]: Mongo docs mention that changes are atomic at a specific document level... but I'll double check.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java (hunk 1 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterFactory.java]: I've made Mongo config properties nullable to match the pattern on other adapters: 4ba74c3889b28dcb70c6b3fb4d9a3048c92f2095

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java (hunk 11 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java:379]: https://docs.mongodb.com/manual/core/write-operations-atomicity/

All write operations are supposed to be atomic at the level of individual documents. I have to assume that `updateOne` is considered one update operation (both the filter part and the change part), so if concurrent operations target the same doc, I suppose one of them will not "find" the changed data, which we catch in the result check (a few lines below).

Having reviewed the docs once more, I believe it might be preferable to use `replaceOne` instead, actually.

Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: Switched to `replaceOne` in 681efd77e4b9657f67632fc35c82476ff7283ab1.

I do not think it makes a material difference in term of atomicity / consistency, but it looks cleaner code-wise.

Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java]: Re: "conditional update". From my POV those pipelines are mostly used for deriving the updated value on the server side. We do not need to do that in our case as we have the complete new doc available at the start of `globalPointerCas()`

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+import org.projectnessie.versioned.persist.adapter.DatabaseAdapterConfig;
+
+@Value.Immutable(lazyhash = true)
+public interface MongoDatabaseAdapterConfig extends DatabaseAdapterConfig {
+
+  @Nullable
+  String getConnectionString();
+
```
Comment (snazy) [inline versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:27]: Missing `MongoDatabaseClient withConnectionString(String connectionString)` +  `MongoDatabaseClient withDatabaseName(String databaseName)`

File: versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    mongo = starter.prepare(mongodConfig);
+    mongo.start();
+
+    assertThat(port).hasPositiveValue();
+
+    createAdapter(
+        "MongoDB",
+        config -> {
+          MongoDatabaseAdapterConfig testConfig =
```
Comment (snazy) [inline versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java:97]: Better move this into `MongoDatabaseAdapter.<init>()` if `config.getClient()` returns `null`.
It's not the greatest approach, but let's keep the functionality the same across all adapter implementations.

Comment (snazy) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java:97]: That way it's also guaranteed that the `MongoDatabaseClient` is closed when the `MongoDatabaseAdpater` is closed and not left behind when a test finishes.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java (hunk 12 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:27]: Why are they required?

File: versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java (hunk 13 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java:97]: The closing of `MongoDatabaseClient` is handled by reference counting already. Will move code for consistency anyway.

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java (hunk 12 above)
Comment (snazy) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:27]: Just completeness - every *DatabaseAdapterConfig sub-type has those.

File: versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java (hunk 13 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/test/java/org/projectnessie/versioned/persist/mongodb/TestMongoDatabaseAdapter.java]: Added default client creation logic to the adapter in a1c226a4c4965e21b3c581d638a8fc302d8a4061

File: versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java (hunk 12 above)
Comment (dimas-b) [reply versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapterConfig.java:27]: Added "with" methods in a1c226a4c4965e21b3c581d638a8fc302d8a4061

**Review by snazy — APPROVED** [review]
Ship it :)
